### ISSUE TITLE
[CLN] Modernize/Organize test suite

### ISSUE THREAD

**Issue by jbrockmendel** [cython/cython#3784]
In writing a test for #3737 I spent some time in the test suite and found more learning curve than I expected.

A few suggestions (for which I'm willing to offer labor):

1) Organize the tests (particularly tests/run) into subdirectories for discoverability, e.g.
   - tests/run/includes/cpython  <- corresponding to Cython/Includes/cpython
   - tests/run/builtins  <- corresponding to python keywords
   - tests/run/options <- e.g. allows_allow_keywords_T295.pyx

2) See if transitioning to use pytest instead of a custom script is viable 

3) Cleanup foo_bugs.txt (ToDo.txtfiles and make sure they all have have GH issues

4) Cleanup other legacy files
   - there's both docs/ and Doc/, looks like the latter may be subsumed by the former?
   - scripts in the root directory may be redundant with scripts in the bin/ directory?  
   - appyveyor/ directory, but AFAICT cython doesnt use appveyor
   - things like Cython/Parser/Grammar that are explicitly unused could go into a "sandbox" directory

5) document what tests/broken/ is.  I'm assuming these are tests that are not passing.  Ideally use something like `pytest.mark.xfail`

**Comment by da-woods** [comment]
> appyveyor/ directory, but AFAICT cython doesnt use appveyor

Cython does use Appveyor to test Cython+Windows (https://ci.appveyor.com/project/cython/cython). It isn't hugely integrated with Github currently which doesn't help with anyone knowing it's there.

(I'm deliberately ignoring the main point here - I suspect there are definitely useful improvements to be made but I don't personally know what they should be)

**Comment by jbrockmendel** [comment]
> major problem here is that pytest has a test discovery phase and a test run phase, whereas for Cython's test modules have two test phases, it's module discovery, compilation test, module test discovery, module test. That's difficult to fit into existing test runners.

A couple of recent numpy PRs made pytest fixtures that did cython compilation steps.  I'll take a look and see if that approach can be adapted for this use case.  If it can, is this something you'd be open to?

Even if its feasible, the process of porting all of the tests either in a) one massive PR or b) incremental PRs could be a PITA.

**Comment by jbrockmendel** [comment]
Looking at setup.py, it looks like the apparently-duplicate bin/foo and foo.py scripts are related to platform support

**Comment by scoder** [comment]
> A couple of recent numpy PRs made pytest fixtures that did cython compilation steps. I'll take a look and see if that approach can be adapted for this use case. If it can, is this something you'd be open to?

Depends. If the whole test discovery fails and aborts because one module cannot compile, then no, that's not acceptable. It would still need to report a failed compilation as some kind of single failure, but continue with all the rest (and then fail in the end). That's what I meant, the compilation itself is already a test in Cython, but needs to run before the module level test discovery (because without compilation, there is no module).

> Even if its feasible, the process of porting all of the tests either in a) one massive PR or b) incremental PRs could be a PITA.

Don't know. Depends on how it's done. pytest supports plugins, which might be a way to adapt the tool to the test suite, instead of vice-versa.

**Comment by mattip** [comment]
I started [a branch](https://github.com/mattip/cython/commit/1a783ca8e6566f2157ff2b3e643093dfbb9b498f) to add windows CI via github actions. The changes so far are quite small, but I ran into these problems
- Apparently on my fork there is some limit to the number of CI jobs. For now I reduced the total to 12 so that the windows jobs run. When I make this into a PR I will revert that bit to see if it is just my fork or some general github limit.
- The job is timing out after 30 minutes. Is that a hung CI run or should the run be parallelized somehow? The [last thing it prints](https://github.com/mattip/cython/runs/3184937873?check_suite_focus=true#step:5:14235) is 
  ```
  Entering debug mode. Use h or ? for help. 
  ```

**Comment by scoder** [comment]
> I started [a branch](https://github.com/mattip/cython/commit/1a783ca8e6566f2157ff2b3e643093dfbb9b498f) to add windows CI via github actions. The changes so far are quite small

Nice!

>     * The job is timing out after 30 minutes. Is that a hung CI run or should the run be parallelized somehow? The [last thing it prints](https://github.com/mattip/cython/runs/3184937873?check_suite_focus=true#step:5:14235) is
>       ```
>       Entering debug mode. Use h or ? for help. 
>       ```

It seems to be looking at a "last exit code", so maybe it crashed and then went into some kind of CLI based post-mortem state?

**Comment by mattip** [comment]
hmm. Maybe I should try it locally first. It is nice that github actions can use bash syntax on windows, I wonder if I can replicate that somehow with WSL.

**Comment by scoder** [comment]
> It seems to be looking at a "last exit code", so maybe it crashed and then went into some kind of CLI based post-mortem state?

Erm, no, it actually seems to get killed after 30 minutes. It doesn't hang before that. So – why does it take 30 minutes to run the tests?

**Comment by scoder** [comment]
The `multiprocessing` setup seems to break. It tries to run multiple test shards (which is fine), but then each of them starts over with the full test run.

**Comment by scoder** [comment]
@mattip maybe it's just a limitation in the machine resources? It runs 7 parallel jobs. If you already had to reduce the total number of test matrix configurations, then maybe there's a similar limitation for each machine.

Can you reduce the `-j7` to, say, `-j3`?

**Comment by scoder** [comment]
BTW, maybe you could get more resources if you opened a PR?

**Comment by 0dminnimda** [comment]
> I started [a branch](https://github.com/mattip/cython/commit/1a783ca8e6566f2157ff2b3e643093dfbb9b498f) to add windows CI via github actions. The changes so far are quite small, but I ran into these problems
> 
> * Apparently on my fork there is some limit to the number of CI jobs. For now I reduced the total to 12 so that the windows jobs run. When I make this into a PR I will revert that bit to see if it is just my fork or some general github limit.
> * The job is timing out after 30 minutes. Is that a hung CI run or should the run be parallelized somehow? The [last thing it prints](https://github.com/mattip/cython/runs/3184937873?check_suite_focus=true#step:5:14235) is
>   ```
>   Entering debug mode. Use h or ? for help. 
>   ```

oh, it wouldn't be cool to be late, but I'm also working on this in this [branch](https://github.com/0dminnimda/cython/pull/1), I didn't do any PR, because I didn't want to load CI and be able to cancel its workflow

**Comment by 0dminnimda** [comment]
> I wonder if I can replicate that somehow with WSL.

GitHub bash is a good place to start

**Comment by 0dminnimda** [comment]
> BTW, maybe you could get more resources if you opened a PR?
idk, but it seems that msvc just runs slower than others, it's regular speed is 30 min, compared to some 10-15 min runs with gcc on Linus for example

**Comment by 0dminnimda** [comment]
@mattip, if you want, we could somehow combine our efforts

**Comment by 0dminnimda** [comment]
> I started [a branch](https://github.com/mattip/cython/commit/1a783ca8e6566f2157ff2b3e643093dfbb9b498f) to add windows CI via github actions.

actually i get it working with c, but not c++

**Comment by jbrockmendel** [comment]
FWIW the pandas builds on GH actions regularly has one of the windows ones timing out

**Comment by mattip** [comment]
> GitHub bash is a good place to start

Are you sure that is what CI is using? It would be good to duplicate the CI run locally.

**Comment by 0dminnimda** [comment]
> Are you sure that is what CI is using? It would be good to duplicate the CI run locally.

Nope, they are not, they use windows with Windows Subsystem for Linux [WSLv1]

**Comment by mcepl** [comment]
> Even if its feasible, the process of porting all of the tests either in a) one massive PR or b) incremental PRs could be a PITA.

Just a fly-by comment: pytest can run both unittest and nose test suites, so the gradual conversion is perfectly possible.
